odoo/addons/base/tests/test_ir_translation.py

```diff
@@ -0,0 +1,123 @@
+
+    @staticmethod
+    def _transform(terms, transformations):
+        """ Transform the terms with the provided transformations
+        :param list(str) terms: terms to transform
+        :param list(str) transformations: transformations to apply among upper, lower and spaces
+        :return: transformed terms
+        """
+        transformed_terms = terms
+        if transformations:
+            for transformation in transformations:
+                if transformation == 'upper':
+                    transformed_terms = [term.upper() for term in transformed_terms]
+                elif transformation == 'lower':
+                    transformed_terms = [term.lower() for term in transformed_terms]
+                elif transformation == 'spaces':
+                    transformed_terms = [f'  {term.lower()}  ' for term in transformed_terms]
+                else:
+                    raise NotImplementedError(f'Transformation {transformation} not implemented')
+        return transformed_terms
+
+    @staticmethod
+    def _translate_and_transform(trans_by_lg_code, lang_code, terms, transformations=None):
+        """ Translate and transform terms
+        :param dict trans_by_lg_code: translation mappings (lang_code -> src -> translated_term)
+        :param str lang_code: target language of the translation
+        :param list(str) terms: terms to translate (in en_US contained in trans_by_lg_code as src)
+        :param transformations: transformation to apply (see _transform)
+        """
+        translated_terms = terms
+        if lang_code in trans_by_lg_code:
+            translations = trans_by_lg_code[lang_code]
+            translated_terms = {translations.get(term, term) for term in terms}
+        return TestIrTranslation._transform(translated_terms, transformations)
+
+    def test_guess_language_of_terms(self):
+        """ Test the detection of the language based on terms present in the table ir.translation """
+        Cls = TestIrTranslation
+        IrTranslation = self.env['ir.translation']
+        ResLang = self.env['res.lang']
+        trans_by_lg_code = {
+            'fr_BE': {
+                'Confirm': 'Confirmer',
+                'Cancel': 'Annuler',
+                'Ok': 'Ok',
+                'email': 'e-mail',
+            },
+            'nl_BE': {
+                'Confirm': 'Bevestigen',
+                'Validate': 'Bevestigen',  # Twice "Bevestigen" to test that it is counted only once
+                'Cancel': 'Annuleren',
+                'Ok': 'Ok',
+                'email': 'e-mail',
+            }
+        }
+        # Activate tested language
+        for lg_code in trans_by_lg_code:
+            lang = ResLang.search([('code', '=', lg_code), ('active', '!=', 1)])
+            if lang:
+                lang.write({'active': 1})
+        # Remove all model translation to avoid interference with the test
+        self.cr.execute("DELETE FROM ir_translation WHERE type='model'")
```

Comment:
I suppose you can't do an unlink because of constraints ?

Reply:
Actually, I didn't test but won't we get worse performance with unlink (select + delete + maybe other methods that will be called).

Reply:
for a test it's ok, and it's always better to use the ORM when we can

Reply:
Thanks for the guideline, it is updated.